bee-signing/src/ternary/scheme.rs

```diff
@@ -0,0 +1,270 @@
+    /// let message = "CHXHLHQLOPYP9NSUXTMWWABIBSBLUFXFRNWOZXJPVJPBCIDI99YBSCFYILCHPXHTSEYSYWIGQFERCRVDD";
+    /// let message_trits = TryteBuf::try_from_str(message).unwrap().as_trits().encode::<T1B1Buf>();
+    /// let signature = private_key.sign(&message_trits).unwrap();
+    /// ```
+    fn sign(&mut self, message: &Trits<T1B1>) -> Result<Self::Signature, Self::Error>;
+}
+
+/// A ternary public key.
+pub trait PublicKey {
+    /// Matching signature type.
+    type Signature: Signature;
+    /// Errors occuring while handling public keys.
+    type Error;
+
+    /// Verifies a signature for a given message.
+    ///
+    /// # Arguments
+    ///
+    /// * `message`     A slice that holds a message to verify a signature for.
+    /// * `signature`   The signature to verify.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bee_crypto::ternary::sponge::Kerl;
+    /// # use bee_signing::ternary::{
+    ///     seed::Seed,
+    ///     wots::{WotsSecurityLevel, WotsSpongePrivateKeyGeneratorBuilder},
+    ///     PrivateKeyGenerator
+    /// };
+    /// # use bee_signing::ternary::PrivateKey;
+    /// use bee_signing::ternary::PublicKey;
+    /// # use bee_ternary::{
+    ///     T1B1Buf,
+    ///     TryteBuf,
+    /// };
+    /// use std::str::FromStr;
+    ///
+    /// # let seed = Seed::from_str("AVXX9XWUSUVKUTWXKTBG9BJVBTZSAISBILKJNVWUHOQNYDMQWXNUCLTTOZGTTLLIYDXXJJGJSEOKVOSSZ").unwrap();
+    /// # let private_key_generator = WotsSpongePrivateKeyGeneratorBuilder::<Kerl>::default()
+    ///     .security_level(WotsSecurityLevel::Medium)
+    ///     .build()
+    ///     .unwrap();
+    /// # let mut private_key = private_key_generator.generate_from_seed(&seed, 0).unwrap();
+    /// # let message = "CHXHLHQLOPYP9NSUXTMWWABIBSBLUFXFRNWOZXJPVJPBCIDI99YBSCFYILCHPXHTSEYSYWIGQFERCRVDD";
+    /// # let message_trits = TryteBuf::try_from_str(message).unwrap().as_trits().encode::<T1B1Buf>();
+    /// # let public_key = private_key.generate_public_key().unwrap();
+    /// # let signature = private_key.sign(&message_trits).unwrap();
+    /// let valid = public_key.verify(&message_trits, &signature).unwrap();
+    /// ```
+    fn verify(&self, message: &Trits<T1B1>, signature: &Self::Signature) -> Result<bool, Self::Error>;
+
+    /// Returns the size of the public key.
+    fn size(&self) -> usize;
+
+    /// Creates a public key from trits.
+    fn from_trits(buf: TritBuf<T1B1Buf>) -> Result<Self, Self::Error>
+    where
+        Self: std::marker::Sized;
+
+    /// Interprets the public key as trits.
+    fn as_trits(&self) -> &Trits<T1B1>;
+}
+
+/// A ternary signature.
+pub trait Signature {
+    /// Errors occuring while handling public keys.
+    type Error;
+
+    /// Returns the size of the signature.
+    fn size(&self) -> usize;
+
+    /// Creates a signature from trits.
+    fn from_trits(buf: TritBuf<T1B1Buf>) -> Result<Self, Self::Error>
+    where
+        Self: std::marker::Sized;
```

Comment:
`Sized` is in the prelude so no `std::marker::` is required.

Reply:
Done